### PR TITLE
[ros2bag test_record] Gets rid of time.sleep and move to using command.wait_for_output

### LINKED ISSUE #454: ros2bag QOS test uses several sleeps on windows
Test is "reliably" failing on windows:
https://ci.ros2.org/view/nightly/job/nightly_win_rel/1605/testReport/ros2bag.test/test_record_qos_profiles/test_record_qos_profiles/

```
FAIL: test_record_qos_profiles.TestRos2BagRecord.test_incomplete_qos_profile
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python38\lib\shutil.py", line 613, in _rmtree_unsafe
    os.unlink(fullname)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\tmphur3j5pi\\ros2bag_test_incomplete_0.db3'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\python38\lib\tempfile.py", line 803, in onerror
    _os.unlink(path)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\ContainerAdministrator\\AppData\\Local\\Temp\\tmphur3j5pi\\ros2bag_test_incomplete_0.db3'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ci\ws\src\ros2\rosbag2\ros2bag\test\test_record_qos_profiles.py", line 87, in test_incomplete_qos_profile
    assert not matches, print('ros2bag CLI did not produce the expected output')
```

@emersonknapp do you have an idea?

### PR DESCRIPTION
**Pull request by jaisontj**
Tries to fix #454 

Looks to me that the original problem described at #454 was probably addressed by #466 
However, before #466, there was #462 which moved from using different temp directory per test to using one temp directory for all the tests - this PR also introduced the `time.sleep` within each test and also one within the cleanup block (which deletes the temp directory)

In this PR:
- I've moved from using `time.sleep` within each tests to using `bag_command.wait_for_output` which essentially waits for the given timeout number or seconds for the expected output. This waiting needs to be done one way or another to give the process enough time to execute and print whatever it needs to.
- I've also improved on the test conditions. For eg: `test_qos_simple` and `test_incomplete_qos_profile` were only testing if the `bag_command.output` was NOT of the format `[ERROR] [ros2bag]`. This causes a bug where the test will pass even if bag_command.output is empty ~ which can happen if the process is exited immediately i.e replace `time.sleep(3)` with a `pass`

### CODE REVIEW

**Comment by jaisontj** [comment]
Each of the tests can also be reduced to:

```python
 def test_nonexistent_qos_profile(self):
        profile_path = PROFILE_PATH / 'foobar.yaml'
        output_path = Path(self.tmpdir.name) / 'ros2bag_test_nonexistent'
        arguments = ['record', '-a', '--qos-profile-overrides-path', profile_path.as_posix(),
                     '--output', output_path.as_posix()]
        expected_string_regex = re.compile(
            r'ros2 bag record: error: argument --qos-profile-overrides-path: can\'t open')
        with self.launch_bag_command(arguments=arguments) as bag_command:
            condition_result = bag_command.wait_for_output(
                condition=lambda output: expected_string_regex.search(output) is not None,
                timeout=5)
            assert condition_result, print('ros2bag CLI did not produce the expected output')
        bag_command.wait_for_shutdown(timeout=5)
        assert bag_command.terminated
        assert bag_command.exit_code != launch_testing.asserts.EXIT_OK
```

However, this can have an edge case where `wait_for_command` gets called after the expected string is printed out.

File: ros2bag/test/test_record_qos_profiles.py [hunk 1]
```diff
@@ -81,38 +80,47 @@ def test_qos_simple(self):
         output_path = Path(self.tmpdir.name) / 'ros2bag_test_basic'
         arguments = ['record', '-a', '--qos-profile-overrides-path', profile_path.as_posix(),
                      '--output', output_path.as_posix()]
+        expected_string_regex = re.compile(
+            r'\[rosbag2_storage]: Opened database .* for READ_WRITE')
         with self.launch_bag_command(arguments=arguments) as bag_command:
-            time.sleep(3)
+            bag_command.wait_for_output(
+                condition=lambda output: expected_string_regex.search(output) is not None,
+                timeout=5)
```
Comment (dabonnie) [inline ros2bag/test/test_record_qos_profiles.py:88]: Defining the timeout would be nice since it's reused (no magic numbers).

File: ros2bag/test/test_record_qos_profiles.py [hunk 2]
```diff
@@ -81,38 +80,47 @@ def test_qos_simple(self):
...
         bag_command.wait_for_shutdown(timeout=5)
         assert bag_command.terminated
-        expected_string_regex = re.compile(ERROR_STRING)
         matches = expected_string_regex.search(bag_command.output)
-        assert not matches, print('ros2bag CLI did not produce the expected output')
+        assert matches, print('ros2bag CLI did not produce the expected output')
```
Comment (dabonnie) [inline ros2bag/test/test_record_qos_profiles.py:92]: In each case can you also print the error output?

**Review by dabonnie — CHANGES_REQUESTED** [review]
Were you able to run the tests repeatedly by any chance?

**Comment by jaisontj** [comment]
> Were you able to run the tests repeatedly by any chance?

I've been testing on windows using: `colcon test --packages-select ros2bag --retest-until-fail 25 --merge-install`

**Comment by jaisontj** [comment]
Looks like the build failure on `Rpr__rosbag2__ubuntu_focal_amd64` is unrelated to this PR as the associated tests have passed: http://build.ros2.org/job/Rpr__rosbag2__ubuntu_focal_amd64/206/console#console-section-5

**Comment by jaisontj** [comment]
Running on CI: 

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12377/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7346/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10089/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12289/)

**Comment by jaisontj** [comment]
CI On windows with --retest-until-fail 25

* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12294/)

**Comment by jaisontj** [comment]
Do note that I was unable to get rid  of the `time.sleep` [here](https://github.com/ros2/rosbag2/blob/6e9d96659bb571d12a0d5b9ebdb293a884bed5e0/ros2bag/test/test_record_qos_profiles.py#L79) as this hack seems to be required.

**Review by dabonnie — APPROVED** [review]

**Review by emersonknapp — APPROVED** [review]
LGTM!

**Review by jikawa-az — APPROVED** [review]
